src/transform/src/dataflow.rs

```diff
@@ -34,6 +37,131 @@ pub fn optimize_dataflow(dataflow: &mut DataflowDesc) {
     monotonic::optimize_dataflow_monotonic(dataflow);
 }
 
+/// Inline views used in one other view, and in no exported objects.
+fn inline_views(dataflow: &mut DataflowDesc) {
+    // We cannot inline anything whose `BuildDesc::id` appears in either the
+    // `index_exports` or `sink_exports` of `dataflow`, because we lose our
+    // ability to name it.
+
+    // A view can / should be in-lined in another view if it is only used by
+    // one subsequent view. If there are two distinct views that have not
+    // themselves been merged, then too bad and it doesn't get inlined.
+
+    // Starting from the *last* object to build, walk backwards and inline
+    // any view (not index, because who knows wtf that is) that is neither
+    // referenced by a `index_exports` nor `sink_exports` nor more than two
+    // remaining objects to build.
+
+    for index in (0..dataflow.objects_to_build.len()).rev() {
+        // Capture the name used by others to reference this view.
+        let global_id = dataflow.objects_to_build[index].id;
+        // Determine if any exports directly reference this view.
+        let mut occurs_in_export = false;
+        for (_gid, sink_desc) in dataflow.sink_exports.iter() {
+            if sink_desc.from.0 == global_id {
+                occurs_in_export = true;
+            }
+        }
+        for (_, index_desc, _) in dataflow.index_exports.iter() {
+            if index_desc.on_id == global_id {
+                occurs_in_export = true;
+            }
+        }
+        // Count the number of subsequent views that reference this view.
+        let mut occurrences_in_later_views = Vec::new();
+        for other in (index + 1)..dataflow.objects_to_build.len() {
+            if dataflow.objects_to_build[other]
+                .relation_expr
+                .as_ref()
+                .global_uses()
+                .contains(&global_id)
+            {
+                occurrences_in_later_views.push(other);
+            }
+        }
+        // Inline if the view is referenced in one view and no exports.
+        if !occurs_in_export && occurrences_in_later_views.len() == 1 {
+            // Test that we are not attempting to inline an index.
+            // TODO: Figure out what that would mean and why indexes
```

Comment:
Would it make sense to annotate each `object_to_build` with arrangement keys, sort of like `ArrangeBy` has? Like, it would be an empty vector for each thing we don't care to arrange, but could have specific `Vec<ScalarExpr>` keys for arrangements we want to be sure are constructed?